examples/dao/assets/dao-app-approval.py

```diff
@@ -0,0 +1,471 @@
+                App.localPut(Int(0), Bytes("from"), Txn.application_args[9]),
+                App.localPut(Int(0), Bytes("recipient"), Txn.application_args[10]),
+                App.localPut(Int(0), Bytes("amount"), Btoi(Txn.application_args[11])),
+            ])],
+            [App.localGet(Int(0), Bytes("type")) == Int(2), Seq([
+                App.localPut(Int(0), Bytes("from"), Txn.application_args[9]),
+                App.localPut(Int(0), Bytes("asa_id"), Btoi(Txn.application_args[10])),
+                App.localPut(Int(0), Bytes("recipient"), Txn.application_args[11]),
+                App.localPut(Int(0), Bytes("amount"), Btoi(Txn.application_args[12])),
+            ])],
+            [App.localGet(Int(0), Bytes("type")) == Int(3), Seq([
+                App.localPut(Int(0), Bytes("msg"), Txn.application_args[9])
+            ])],
+        ),
+        # Finally record the following special attributes: id (the TxID field) and executed = 0
+        App.localPut(Int(0), Bytes("id"), Txn.tx_id()),
+        App.localPut(Int(0), Bytes("executed"), Int(0)),
+        Return(Int(1))
+    ])
+
+    # sender.deposit
+    deposit = App.localGet(Int(0), Bytes("deposit"))
+
+    # Records gov tokens deposited by user (sender)
+    deposit_vote_token = Seq([
+        verify_deposit(Int(1), App.globalGet(deposit_lsig)),
+        # Sender.deposit += amount
+        App.localPut(Int(0), Bytes("deposit"), deposit + Gtxn[1].asset_amount()),
+        Return(Int(1))
+    ])
+
+    # This is used to unlock the deposit and withdraw tokens back to the user.
+    # To protect against double vote, user can only withdraw the deposit after the
+    # latest voting he participated has ended.
+    withdraw_vote_deposit = Seq([
+        Assert(And(
+            Global.latest_timestamp() > App.localGet(Int(0), Bytes("deposit_lock")),
+            # fees must be paid by tx0 (voter)
+            Gtxn[1].fee() == Int(0)
+        )),
+        App.localPut(Int(0), Bytes("deposit"), deposit - Gtxn[1].asset_amount()),
+        Return(Int(1))
+    ])
+
+    # p_<proposal> is a concatenation of p_ with the proposal address to avoid some weird attacks.
+    byte_p_proposal = Concat(Bytes("p_"), Txn.accounts[1])
+    p_proposal = App.localGetEx(Int(0), Int(0), byte_p_proposal)  # value = proposal.id when a user voted or 0
+    yes = Bytes("yes")
+    no = Bytes("no")
+    abstain = Bytes("abstain")
+
+    # Register user votes in proposal_lsig by saving Sender.p_<proposal>. Arguments:
+    # * proposal : lsig account address with the proposal record (provided as the first external account).
+    # * vote (bytes): abstain, yes, no
+    register_vote = Seq([
+        p_proposal,
+        Assert(
+            And(
+                Global.group_size() == Int(1),
+                # voting_start <= now <= voting_end
+                # voting_start <= Global.latest_timestamp(),
+                # Global.latest_timestamp() <= voting_end,
```

Comment:
where do we have this checks? They are essential. Voting is only possible when `voting_start <= now <= voting_end`

Reply:
yes, i was having some issue with block timestamp in a development network. Will test & uncomment

Reply:
OK

Reply:
ping - we need to add the voting_start and voting_end checks.

Reply:
done, i increased the buffer by 1min `{ voting start = now + 1min, voting start = now + 3min }`.